File: syllabus-1/day-4/PAC6.cpp
#include <bits/stdc++.h>
using namespace std;

#define endl "\n"
#define ll long long

void solve(){
    int n,r,x,y;
    ll m;
    cin>>n>>m>>r>>x>>y;
    x--;
    y--;
    
    vector<vector<pair<int,int>>> arr(n);
    for(int i=0;i<r;++i){
        int a,b,c;
        cin>>a>>b>>c;
        a--;    b--;
        
        arr[a].push_back({b,c});
        arr[b].push_back({a,c});
    }
    
    vector<ll> costs(n,-1);
    vector<bool> visited(n,0);
    
    priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> q;
    q.push({0,x});
    costs[x] = 0;
    
    while(q.empty() == false){
        pair<int,int> temp = q.top();
        q.pop();
        int node = temp.second;
        int val = temp.first;
        
        if(visited[node] == 1) continue;
        
        // cout<<node + 1<<endl;
        
        visited[node] = 1;
        for(auto i: arr[node]){
            if(visited[i.first] == 0){
                if(costs[i.first] == -1){
                    costs[i.first] = i.second + val;
                }else{
                    costs[i.first] = min(costs[i.first],(ll)i.second + val);
                }
                q.push({costs[i.first],i.first});
            }
        }
        
        // for(ll i: costs){
        //     cout<<i<<" ";
        // }cout<<endl;
    }
    
    if(costs[y] > m){
        cout<<"died"<<endl;
        return;
    }
    cout<<"alive "<<costs[y]<<endl;
}

int main(){
    
    int t = 1;
    // cin>>t;
    while(t--){
        solve();
    }
    
    return 0;
}

File: syllabus-1/day-4/PAC6.py
import heapq


def dfs(st, arr, costs):
    visited = [0 for i in range(len(arr))]

    q = []
    heapq.heappush(q, (0, st))
    costs[st] = 0

    while(len(q) > 0):
        temp = heapq.heappop(q)
        val = temp[0]
        node = temp[1]

        if(visited[node] == 1):
            continue

        visited[node] = 1
        for i in arr[node]:
            if(visited[i[0]] != 0):
                continue

            if(costs[i[0]] == -1):
                costs[i[0]] = i[1] + val
            else:
                costs[i[0]] = min(costs[i[0]], i[1] + val)
            heapq.heappush(q, (costs[i[0]], i[0]))


def solve():
    (n, m, r, x, y) = map(int, input().split())
    x -= 1
    y -= 1

    arr = [[] for i in range(n)]
    for i in range(r):
        (a, b, c) = map(int, input().split())
        a -= 1
        b -= 1

        arr[a].append((b, c))
        arr[b].append((a, c))

    costs = [-1 for i in range(n)]
    dfs(x, arr, costs)

    if(costs[y] > m):
        print("died")
        return
    print("alive " + str(costs[y]))


solve()
